<a href="https://colab.research.google.com/github/rilschultz/CS167Notes/blob/main/Project1_starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project \#1 Starter Code: 
Your first step is to decide which dataset you would like to work with for this project. 

For this project, you will get to choose what dataset you use. The ones we've used in class (other than Iris/Titanic as they're pretty small) are all fair game (wine quality, breast cancer, spotify). 

Here are some other possibilities--Kaggle has many many more. It might be a good idea to double check with me before grabbing a dataset that's not listed here:
- 🥑 [Avocado Prices](https://www.kaggle.com/datasets/neuromusic/avocado-prices)
- 🎟️ [Admission to University](https://www.kaggle.com/datasets/akshaydattatraykhare/data-for-admission-in-the-university)
- 🏎️ [Speed Dating](https://www.kaggle.com/datasets/whenamancodes/speed-dating)
- 🎓 [World University Ranking](https://www.kaggle.com/datasets/whenamancodes/world-university-ranking-2022-2023)
- 📸 [Top Instagram Influencers](https://www.kaggle.com/datasets/whenamancodes/top-200-influencers-crushing-on-instagram)
- 🌦️ [Weather Prediction](https://www.kaggle.com/datasets/thedevastator/weather-prediction)
- 🤑 [Student Monthly Expenses](https://www.kaggle.com/datasets/shariful07/nice-work-thanks-for-share)
- 🍺 [Student Alcohol Consumption](https://www.kaggle.com/datasets/uciml/student-alcohol-consumption)
- 🏫 [College Score Card](https://www.kaggle.com/datasets/thedevastator/u-s-department-of-education-college-scorecard-da)
- 🎮 [IMDB Video Games](https://www.kaggle.com/datasets/muhammadadiltalay/imdb-video-games)
- 🎵 [Spotify Recommendation](https://www.kaggle.com/datasets/bricevergnou/spotify-recommendation)

# Self-Evaluation:

| **Description/Writing** | **I think I deserve** |
|-------------------------|-----------------------|
| 1: Problem              | /2                    |
| 2: Data Prep            | /2                    |
| 3: Research             | /10                   |
| 4: Analysis             | /10                   |
| 5: Bumps in the Road.   | /1                    |
| Total                   | /25                   |

# 1. Problem:
State the problem you are trying to solve with this machine learning experiment. Include a description of the data, where you got the data, and what you're trying to predict..

For Project 1 I plan to use the machine learning algoritms we learned on the addmitted student data set. With these I will see which factors are most important to univerisities when accepting a prospective student. Because this application process can be stressful I want to be able to show what aspects are important as well as give a good estimate based on the algorithm to see if a student would be accepted or not given their statistics.

# 2. Data Preparation
Explain your data preparation. What did you have to do to get your data in shape for your experiments? Why are you certain that you data is clean and prepared for use in your algorithms?

First step is to import my drive and upload the data set

In [12]:
# load and preapre your data here:
from google.colab import drive
drive.mount('/content/drive')

import pandas as pd 
path = '/content/drive/MyDrive/datasets/adm-data.csv' 
admData = pd.read_csv(path) 
admData.head()



Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0           1        337          118                  4  4.5   4.5  9.65   
1           2        324          107                  4  4.0   4.5  8.87   
2           3        316          104                  3  3.0   3.5  8.00   
3           4        322          110                  3  3.5   2.5  8.67   
4           5        314          103                  2  2.0   3.0  8.21   

   Research  Chance of Admit   
0         1              0.92  
1         1              0.76  
2         1              0.72  
3         1              0.80  
4         0              0.65

The next step is to check if there is any data that is nan and would need to be removed

In [18]:
admData.isna().any()


Serial No.           False
GRE Score            False
TOEFL Score          False
University Rating    False
SOP                  False
LOR                  False
CGPA                 False
Research             False
Chance of Admit      False
dtype: bool

Since there is no NA values all being false and all values are numeric we can start to normalize the data.

In [24]:
import sklearn
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn import neighbors
#This is selecting which values I want to be my predictors and which one I want to be my target variable
target = 'Chance of Admit '
predictors = admData.columns.drop(target)
train_data, test_data, train_sln, test_sln = train_test_split(admData[predictors], admData[target])

#Next step is to normalize the data to not have values such as GRE and TOEFL scores outweigh other values in the dataset
#Normalize both the train and test data keep solutions unnormalized
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(train_data)
train_data_normalized = scaler.transform(train_data)
test_data_normalized = scaler.transform(test_data)
train_data_normalized

array([[-1.4985191 ,  0.92758573,  1.34361086, ...,  1.72710582,
         1.46167646,  0.91678285],
       [ 0.42356386,  0.33134146,  0.54033613, ...,  0.60560853,
         0.37135891,  0.91678285],
       [ 1.36343702,  0.41651921,  0.37968119, ...,  1.72710582,
         0.06216437,  0.91678285],
       ...,
       [-0.98201222,  0.41651921,  1.18295591, ...,  1.16635718,
         1.23384891,  0.91678285],
       [-0.2199529 ,  0.50169696,  0.86164602, ...,  1.16635718,
         1.02229476,  0.91678285],
       [-1.60012701,  0.67205247, -0.26293859, ...,  0.60560853,
        -0.32839714,  0.91678285]])

# 3. Research

Put your code and your experiments here.

For full credit, your experiments should include the following models:
- __k Nearest Neighbors__
- __weighted k Nearest Neighbors__
- __decision tree__
- __random forest__ 

You need to demonstrate the performance of these models on both __normalized__ and __non-normalized__ data. 

For each of the above models, you need to include a __graph of at least 1 tuned parameter__. 


In [46]:
# code goes here... don't forget to include graphs. Professor Moore loves graphs.
import sklearn
import pandas
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn import neighbors

knn = neighbors.KNeighborsRegressor(n_neighbors=35)

# fit (train) the model to the data
knn.fit(train_data_normalized, train_sln)

# use the trained model to get predictions from the normalized data
preds = knn.predict(test_data_normalized)
# use a metric to see how good our predictions are
print("R2: ", metrics.r2_score(test_sln, preds))
print("MAE: ", metrics.mean_absolute_error(test_sln, preds))
print("MSE: ", metrics.mean_squared_error(test_sln, preds))

R2:  0.8310813999812451
MAE:  0.03796857142857143
MSE:  0.0027251485714285713


In [33]:
knn = neighbors.KNeighborsRegressor(n_neighbors=30, weights = 'distance')

# fit (train) the model to the data
knn.fit(train_data_normalized, train_sln)

# use the trained model to get predictions from the normalized data
preds = knn.predict(test_data_normalized)
# use a metric to see how good our predictions are
print("R2: ", metrics.r2_score(test_sln, preds))
print("MAE: ", metrics.mean_absolute_error(test_sln, preds))
print("MSE: ", metrics.mean_squared_error(test_sln, preds))

R2:  0.8361288159400584
MAE:  0.03717049462303334
MSE:  0.002643719064032473


#4. Analysis

What did you discover? What insights/recommendations do you have? What did you find that was interesting? Which model was your best model, which models didn't work well? Why do you think this is? In general, I want a discussion of your experiment, the results, and what they mean.

This chart might help organize your results:

| **Model**     | **Non-Normalized** | **Normalized** | **Parameter Tuned** |
|---------------|--------------------|----------------|---------------------|
| kNN           |       R2=62%       |   R2=83%       |num neighbors        |
| w-kNN         |                    |                |                     |
| decision tree |                    |                |                     |
| random forest |                    |                |                     |

*your answer here*

# 5. Bumps in the Road
What challenges did you encounter? How did you overcome these challenges?

First was importing the data it took me a couple tries and worked, next was seperating the data there was a hidden space after some columns names so it was harder to type those out.